scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala

```diff
@@ -422,8 +422,21 @@ class ScalametaParser(input: Input, dialect: Dialect) { parser =>
     val pos = TokenStreamPosition(startTokenPos, endTokenPos + 1)
     result.withOrigin(Origin.Parsed(input, dialect, pos)).asInstanceOf[T]
   }
+
+  def atPosE[E, T <: Tree](start: Pos, end: Pos)(body: => Either[E, T]): Either[E, T] = {
```

Comment:
should we just use `Try` instead of `Either`? do you envision a use case when something other than an exception would be used?

Reply:
Check `simpleExpr0` where I create `Left` in one specific case when parsing didn't fall into any match case.
It is hardly doable to get the same results with Try without some weird stuff.
Because if e.g. parsing of `case Underscore()` fails I still want it to throw exception that was raised and not return `Left(...)`

Try could definitely replace Either but how to use that in `simpleExpr0` would be problematic. I don't want to wrap everything in Try because that would change a logic.

Ideally I would like something like:
`def atPosE[F, T <: Tree](start: Pos, end: Pos)(body: => F[T]): F[T]`

but Either is not right biased in 2.11 - what can you do? :/

To sum up what I want:
- in `simpleExpr0` if fall into `case _ =>` notify parent method about this! 
- in `simpleExpr0` if exception is thrown -> keep it throwing to parent methods
- parent of `simpleExpr0` if notified about `case _ =>` continue parsing assuming unary_op is actually defined object name.

Reply:
Panie Krzyśku, my bad, I should have been clearer.

I am _not_ suggesting to wrap everything in `Try`. I am simply suggesting to replace `Either` with `Try`, `Right` with `Success`, and `Left` with `Failure`. In cases where you currently throw an exception, you keep that and not wrap in `Failure`.

Would that work?

Reply:
Total respect for perfect conversion "Krzysztof" -> "Krzyśku" even with ś!

Yes using explicitly `Left` and `Right` and `.get` where I want walue or want to throw exception will work.
It might be misleading that a function returns `Try[X]` and can also additionally throw exception but if already 2 people don't like Either here there must be something in it :) Will change to Try :+1:

Reply:
> Total respect for perfect conversion "Krzysiek" -> "Krzyśku" even with ś!

Well, I grew up w Litwie, so I still have some rudimentary knowledge języka polskiego. Clearly, I had to look up the nasal "e" and the accented soft "sh" as my keyboard doesn't have them. :) Fun exercise.

> Yes using explicitly `Left` and `Right` and `.get` where I want walue or want to throw exception will work.
> It might be misleading that a function returns `Try[X]` and can also additionally throw exception but if already 2 people don't like Either here there must be something in it :) Will change to Try 👍

You're right; if you believe that it will be confusing to use `Try` which lets some exceptions sneak through, let's not use it. We use it routinely, so I didn't think twice about that.

Reply:
changed to `Try`